ATTESTATION_VERIFICATION_PROPOSAL.md

```diff
@@ -0,0 +1,253 @@
+2. **Metadata:** Metadata is used to describe software and the build environment. Provenance (origin) data, SBOMs, and vulnerability scan reports are the essential set of metadata required to assess security risks for software.
+3. **Attestations:**  Authenticated metadata is used to attest to the integrity of a software system. Both custom and standardized metadata can be converted into attestations.
+4. **Policies:** Policies check and enforce organization standards. Policies should be automatically enforced prior to deployment and via runtime scanning. 
+5. **Payload:** Payload in this case refers to the data in the artifact.
+ 
+# Motivation
+[motivation]: #motivation
+
+- Adds ability to verify attestations and do conditional checking on it.
+- Verification of multiple attestations attached to the image.
+
+# Proposal
+
+Currently, we support image verification by passing the `images` variable from the context to `/checkimages` endpoint which returns a list of JSONPatch compatible list to do mutation on digests.
+
+Current request format:
+```json
+{
+  "images": {
+    "containers": {
+      "tomcat": {
+        "registry": "https://ghcr.io",
+        "path": "tomcat",
+        "name": "tomcat",
+        "tag": "9",
+        "jsonPointer": "spec/container/0/image"
+      }
+    },
+    "initContainers": {
+      "vault": {
+        "registry": "https://ghcr.io",
+        "path": "vault",
+        "name": "vault",
+        "tag": "v3",
+        "jsonPointer": "spec/initContainer/0/image"
+      }
+    }
+  }
+}
+```
+We are passing the `images` variable which has `containers`, `initContainers` and `ephimeralContainers` in it, containing information of all the images used.
+
+Current Response Format:
+```json
+{
+  "verified": true,
+  "message": "...",
+  "results": [
+      {
+         "name": "container1",
+         "path":  "/spec/containers/0",
+         "image":  "ghcr.io/kyverno/test-verify-image@sha256:b31bfb4d0213f254d361e0079deaaebefa4f82ba7aa76ef82e90b4935ad5b105"
+      } 
+  ]
+}
+```
+It returns a verified `boolean` that indicated whether verification was successful or not, `message` which is an optional field, and `results` which is a list of JSONPatch compatible objects containing entries for every image.
+
+We should add an array to the `requestData` called `attestationType` as follows
+```json
+{
+  "images": {
+    "containers": {
+      "tomcat": {
+        "registry": "https://ghcr.io",
+        "path": "tomcat",
+        "name": "tomcat",
+        "tag": "9",
+        "jsonPointer": "spec/container/0/image"
+      }
+    },
+    "initContainers": {
+      "vault": {
+        "registry": "https://ghcr.io",
+        "path": "vault",
+        "name": "vault",
+        "tag": "v3",
+        "jsonPointer": "spec/initContainer/0/image"
+      }
+    }
+  },
+  "attestationType": [
+    "application/vnd.cyclonedx",
+    "sbom/cyclone-dx",
+    "application/sarif+json"
+  ]
```

Comment:
>For every image we will check whether it has any or all of the attestations attached to it. 

This could be less efficient if a user only wants to verify a single type for a given image.

How about we request a variable path and its desired attestation types? For example:

```
{
    "attestationType": [
        {
            "path": "images.containers.tomcat",
            "types": [
                "application/vnd.cyclonedx",
                "sbom/cyclone-dx"
            ]
        },
        {
            "path": "images.initContainers.value",
            "types": [
                "application/sarif+json"
            ]
        }
    ]
}
```

Reply:
@realshuting can we do something like this for attestations?
```yaml
context:
- name: response
  apiCall:
    method: POST
    data:
    - key: images
      value: "{{ images }}"
    - key: attestations
      value:
        containers:
          - name: tomcat
            types: |-
            - application/vnd.cyclonedx
            - sbom/cyclone-dx"
          - name: vault
            types: |-
            - application/vnd.cyclonedx
        initContainers:
          - name: vault
            types: |-
            - application/sarif+json
    service:
      url: https://svc.kyverno-notation-aws/checkimages
      caBundle: |-
...
```

In json, it would be 
```json
{
  "images": {
    "containers": {
      "tomcat": {
        "registry": "https://ghcr.io",
        "path": "tomcat",
        "name": "tomcat",
        "tag": "9",
        "jsonPointer": "spec/container/0/image"
      }
    },
    "initContainers": {
      "vault": {
        "registry": "https://ghcr.io",
        "path": "vault",
        "name": "vault",
        "tag": "v3",
        "jsonPointer": "spec/initContainer/0/image"
      }
    }
  },
  "attestations": {
    "containers": [
      {
        "name": "tomcat",
        "types": [
          "application/vnd.cyclonedx",
          "sbom/cyclone-dx"
        ]
      },
       {
        "name": "vault",
        "types": [
          "application/vnd.cyclonedx",
        ]
      }
    ],
    "initContainers": [
      {
        "name": "vault",
        "types": [
          "application/sarif+json"
        ]
      }
    ]
  }
}
```

I think this structure will make it easier to get the image reference from `ImageInfo` struct.
For containers in `attestations` object, `ImageInfos.Containers[element.name].String()`

Reply:
Can we further simplify the `attestations` value? In most cases, it will apply to all. In some cases, there may be image specific attestations. And, I don't think we need the "initContainer", "container", ... grouping. So, maybe just a list of entries that has the image references (or "*" for all) and list of attestation types to return?